### DISCUSSION TITLE
SSIM loss difference

### DISCUSSION THREAD

**Discussion by blaz-r** [kornia/kornia#2626] [Potential Issue] [answered]
Hello.
I have a question regarding the [ssim_loss](https://kornia.readthedocs.io/en/latest/_modules/kornia/losses/ssim.html) from your library. It seems to produce different results than the two other implementations that I have tried, one from [RIAD model](https://github.com/taikiinoue45/RIAD/blob/main/riad/criterions/ssim.py) and other from [DRAEM](https://github.com/VitjanZ/DRAEM/blob/main/loss.py). 

The results of the other two match between each-other, while the kornia ssim_loss is a bit different (even after multiplying by 2).

Following is an example of code:
```python
from kornia.losses import SSIMLoss

draem_loss = SSIM(window_size=11, val_range=1.0) # from DRAEM
riad_loss = SSIMLoss_riad(kernel_size=11) # from RIAD
kornia_loss = SSIMLoss(window_size=11, max_val=1.0)

x = torch.rand(4, 3, 50, 50)
y = torch.rand(4, 3, 50, 50)

print("draem: ", draem_loss(x, y))
print("riad: ", riad_loss(x, y))
print("kornia:", kornia_loss(y, x) * 2)
```
which produces:
```python
draem:  tensor(0.9139)
riad:  tensor(0.9139)
kornia: tensor(0.9896)
```
All produce very similar results in training.

What could be the case here if anyone maybe knows?

**Comment by edgarriba** [comment]
i see that we average in https://github.com/kornia/kornia/blob/master/kornia/losses/ssim.py#L53

**Reply by blaz-r** [reply]
Hey, thanks for the answer.

I don't see how that line would affect this. Even if I change the line in kornia to `loss = 1.0 - ssim_map`, the difference persists.

One thing I did notice when comparing the ssim_map produced by either method is that max value differs:
Draem:
```
ssim_map.max()
tensor(0.8665)
ssim_map.min()
tensor(-0.6337)
```
Kornia
```
ssim_map.max()
tensor(0.6580)
>>>ssim_map.min()
tensor(-0.6337)
```

**Reply by edgarriba** [reply, marked as answer]
then the only difference is the separable convolution? maybe some of the "border" parameters do not match well

**Reply by blaz-r** [reply]
Okay, thanks for the answer. I guess these two things could be the case here.
